### PR TITLE
Homework shared_ptr, weak_ptr, make_shared

### PR DESCRIPTION
**Pull request by Kuba-Wi**
Collaboration with kuchichan

### CODE REVIEW

**Review by MaciejSalamonski — APPROVED** [review]
Imo it look ok +1

**Review by WBigaj — APPROVED** [review]
I have checked your code and it looks good 👍

File: homework/shared_ptr/weak_ptr-ut.cpp [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+TEST(WeakPointerTests, copyAssignmentOperatorShouldMovePointer) {
+    constexpr size_t numberOfUseCount = 1;
+    struct Something {};
+    cs::shared_ptr<Something> shared_something(new Something{});
+    cs::weak_ptr<Something> weak_something{shared_something};
+    cs::weak_ptr<Something> weak_ptr;
+    weak_ptr = weak_something;
+
+    ASSERT_EQ(weak_ptr.use_count(), numberOfUseCount);
+    ASSERT_EQ(weak_something.use_count(), numberOfUseCount);
```
Comment (ziobron) [inline homework/shared_ptr/weak_ptr-ut.cpp:56]: you may also show that before copy use_count was 0. The same in move in a test above.

File: homework/shared_ptr/weak_ptr-ut.cpp [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+TEST(WeakPointerTests, WeakPointerResetShouldSetPtrAndControlBlockAsNullPtr) {
+  int expectedValue{100};
+  auto s_ptr = cs::shared_ptr<int>(new int{20});
+  auto next_ptr = cs::shared_ptr<int>(new int{expectedValue});
+  cs::weak_ptr<int> weakSomething{s_ptr};
+  weakSomething.reset();
```
Comment (ziobron) [inline homework/shared_ptr/weak_ptr-ut.cpp:64]: you should also check if after reset we have use_count == 0 or pointer is expired and it is not after reassignment.

**Review by ziobron — CHANGES_REQUESTED** [review]

File: homework/shared_ptr/weak_ptr.hpp [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    cb_ = ptr.cb_;
+    cb_->increaseWeakRef();
+}
+
+template <typename T>
+weak_ptr<T>::weak_ptr(weak_ptr&& ptr) noexcept {
+    if (this != &ptr) {
+        ptr_ = ptr.ptr_;
+        cb_ = ptr.cb_;
+        cb_->setDeleter([&]() { delete ptr_; });
```
[the remark this answers is not in the document]
Comment (Kuba-Wi) [reply homework/shared_ptr/weak_ptr.hpp:73]: It is needed because otherwise deleter will call delete on old pointer that doesn't point on anything.

File: homework/shared_ptr/weak_ptr.hpp [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    return *this;
+}
+
+template <typename T>
+weak_ptr<T>& weak_ptr<T>::operator=(weak_ptr&& ptr) noexcept {
+    if (&ptr != this) {
+        deletePointers();
+        ptr_ = ptr.ptr_;
+        cb_ = ptr.cb_;
+        cb_->setDeleter([&]() { delete ptr_; });
```
[the remark this answers is not in the document]
Comment (Kuba-Wi) [reply homework/shared_ptr/weak_ptr.hpp:96]: It is needed because otherwise deleter will call delete on old pointer that doesn't point on anything.

File: homework/shared_ptr/shared_ptr.hpp [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    ptr_ = ptr.ptr_;
+    cb_ = ptr.cb_;
+    cb_->increaseSharedRef();
+}
+
+template <typename T>
+shared_ptr<T>::shared_ptr(shared_ptr&& ptr) {
+    ptr_ = ptr.ptr_;
+    cb_ = ptr.cb_;
+    cb_->setDeleter([&]() { delete ptr_; });
```
[the remark this answers is not in the document]
Comment (Kuba-Wi) [reply homework/shared_ptr/shared_ptr.hpp:82]: It is needed because otherwise deleter will call delete on old pointer that doesn't point on anything.

File: homework/shared_ptr/shared_ptr.hpp [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    return *this;
+}
+
+template <typename T>
+shared_ptr<T>& shared_ptr<T>::operator=(shared_ptr&& ptr) {
+    if (&ptr != this) {
+        deletePointers();
+        ptr_ = ptr.ptr_;
+        cb_ = ptr.cb_;
+        cb_->setDeleter([&]() { delete ptr_; });
```
[the remark this answers is not in the document]
Comment (Kuba-Wi) [reply homework/shared_ptr/shared_ptr.hpp:104]: It is needed because otherwise deleter will call delete on old pointer that doesn't point on anything.

File: homework/shared_ptr/shared_ptr.hpp [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    shared_ptr(const shared_ptr& ptr);
+    shared_ptr(const cs::weak_ptr<T>& ptr);
+    shared_ptr(shared_ptr&& ptr);
+    shared_ptr& operator=(const shared_ptr& ptr);
+    shared_ptr& operator=(shared_ptr&& ptr);
+    ~shared_ptr();
+    T* get() const { return ptr_; }
+    T& operator*() const { return *ptr_; }
+    T* operator->() const { return ptr_; }
+    long use_count() const { return static_cast<long>(cb_->getShared()); }
```
[the remark this answers is not in the document]
Comment (Kuba-Wi) [reply homework/shared_ptr/shared_ptr.hpp:28]: Declaration on cppreference:
long use_count() const noexcept;

File: homework/shared_ptr/control_block.hpp [hunk 8]
```diff
@@ -3,31 +3,31 @@
 #include <atomic>
 #include <functional>
 
+template <typename T>
 class ControlBlock {
 public:
-    ControlBlock(std::function<void()> del) : deleter(del) {}
+    ControlBlock() = default;
```
Comment (ziobron) [inline homework/shared_ptr/control_block.hpp:9]: with only default constructor, you don't have a way to provide a custom deleter.

File: homework/shared_ptr/control_block.hpp [hunk 9]
```diff
@@ -3,31 +3,31 @@
...
 
 template <typename T>
-class BlockAndData : public ControlBlock {
+class BlockAndData : public ControlBlock<T> {
 public:
     template <typename... Args>
-    BlockAndData<T>(Args... args) : object_(args...), ControlBlock([]() {}) {}
-    T* getObject() { return &object_; }
+    BlockAndData<T>(Args... args) : object_(args...), ControlBlock<T>() {}
+    const T& getObject() const { return object_; }
```
Comment (ziobron) [inline homework/shared_ptr/control_block.hpp:30]: why not return T*? you cannot modify underlying object with const&. Previous version was ok.

File: homework/shared_ptr/shared_ptr.hpp [hunk 10]
```diff
@@ -14,8 +14,7 @@ namespace cs {
 template <typename T>
 class shared_ptr {
 public:
-    shared_ptr(T* ptr = nullptr)
-        : ptr_(ptr), cb_(new ControlBlock([&]() { delete ptr_; })) {}
+    shared_ptr(T* ptr = nullptr) : ptr_(ptr), cb_(new ControlBlock<T>) {}
```
Comment (ziobron) [inline homework/shared_ptr/shared_ptr.hpp:17]: passing a deleter here was really fine. Bring back the previous version :D

File: homework/shared_ptr/control_block.hpp [hunk 11]
```diff
@@ -3,31 +3,31 @@
...
 class ControlBlock {
 public:
-    ControlBlock(std::function<void()> del) : deleter(del) {}
+    ControlBlock() = default;
     void increaseSharedRef() { shared_refs += 1; }
     void decreaseSharedRef() { shared_refs -= 1; }
     void increaseWeakRef() { weak_refs += 1; }
     void decreaseWeakRef() { weak_refs -= 1; }
     std::atomic<size_t>& getShared() { return shared_refs; }
     std::atomic<size_t>& getWeak() { return weak_refs; }
```
Comment (ziobron) [inline homework/shared_ptr/control_block.hpp]: return size_t not atomic<size_t>

File: homework/shared_ptr/shared_ptr.hpp [hunk 12]
```diff
@@ -38,17 +37,17 @@ class shared_ptr {
     template <typename>
     friend class cs::weak_ptr;
 
-    shared_ptr(BlockAndData<T>* cb) : cb_(cb) { ptr_ = cb->getObject(); };
+    shared_ptr(BlockAndData<T>* cb) : cb_(cb) { ptr_ = new T{cb->getObject()}; };
```
Comment (ziobron) [inline homework/shared_ptr/shared_ptr.hpp:40]: no, you have 2 objects right now. In a control block and a newly created on the heap.

File: homework/shared_ptr/shared_ptr.hpp [hunk 13]
```diff
@@ -110,8 +101,9 @@ shared_ptr<T>& shared_ptr<T>::operator=(shared_ptr&& ptr) {
 
 template <typename T>
 void shared_ptr<T>::reset(T* ptr) {
-    *ptr_ = *ptr;
-    delete ptr;
+    deletePointers();
+    ptr_ = ptr;
+    cb_ = new ControlBlock<T>;
```
Comment (ziobron) [inline homework/shared_ptr/shared_ptr.hpp:106]: and decrement shared refs if ptr_ is nullptr. Add a proper test for it. use_count should return 0 for nulltprs.

File: homework/shared_ptr/shared_ptr.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (ziobron) [reply homework/shared_ptr/shared_ptr.hpp]: ok

File: homework/shared_ptr/shared_ptr.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (ziobron) [reply homework/shared_ptr/shared_ptr.hpp]: No. it will be called on the pointer that was set during the construction of the control block, and it was exactly ptr.ptr_, which you have just copied to ptr_. So it should be ok.

File: homework/shared_ptr/weak_ptr-ut.cpp [hunk 14]
```diff
@@ -34,35 +34,41 @@ TEST(WeakPointerTests, WeakPointerLock) {
...
-    constexpr size_t numberOfUseCount = 1;
+    size_t numberOfUseCount = 0;
     struct Something {};
     cs::shared_ptr<Something> shared_something(new Something{});
     cs::weak_ptr<Something> weak_something{shared_something};
     cs::weak_ptr<Something> weak_ptr;
+    ASSERT_EQ(weak_ptr.use_count(), numberOfUseCount);
     weak_ptr = std::move(weak_something);
-
+    ++numberOfUseCount;
```
Comment (ziobron) [inline homework/shared_ptr/weak_ptr-ut.cpp:44]: I really don't like the incrementation here and in a few other places. It tells me that the first the other use_count has something in common. In fact, they are independent, because they use another share_ptr. I opt for having two separated numberOfUseCount variables with 0 and 1 values here.

File: homework/shared_ptr/weak_ptr-ut.cpp [hunk 15]
```diff
@@ -34,35 +34,41 @@ TEST(WeakPointerTests, WeakPointerLock) {
 }
 
 TEST(WeakPointerTests, moveAssignmentOperatorShouldMovePointer) {
-    constexpr size_t numberOfUseCount = 1;
+    size_t numberOfUseCount = 0;
     struct Something {};
     cs::shared_ptr<Something> shared_something(new Something{});
     cs::weak_ptr<Something> weak_something{shared_something};
     cs::weak_ptr<Something> weak_ptr;
```
Comment (ziobron) [inline homework/shared_ptr/weak_ptr-ut.cpp:41]: please initialize this weak_ptr with another shared_ptr before move. With such empty weak_ptr, some resources may not leak.

File: homework/shared_ptr/weak_ptr-ut.cpp [hunk 16]
```diff
@@ -34,35 +34,41 @@ TEST(WeakPointerTests, WeakPointerLock) {
...
 }
 
-TEST(WeakPointerTests, copyAssignmentOperatorShouldMovePointer) {
-    constexpr size_t numberOfUseCount = 1;
+TEST(WeakPointerTests, copyAssignmentOperatorShouldCopyPointer) {
+    size_t numberOfUseCount = 0;
     struct Something {};
     cs::shared_ptr<Something> shared_something(new Something{});
     cs::weak_ptr<Something> weak_something{shared_something};
     cs::weak_ptr<Something> weak_ptr;
```
Comment (ziobron) [inline homework/shared_ptr/weak_ptr-ut.cpp:53]: please initialize this weak_ptr with another shared_ptr before copy. With such empty weak_ptr, some resources may not leak.

File: homework/shared_ptr/weak_ptr-ut.cpp [hunk 17]
```diff
@@ -34,35 +34,41 @@ TEST(WeakPointerTests, WeakPointerLock) {
...
-  weakSomething.reset();
-  weakSomething = next_ptr;
-  auto locked = weakSomething.lock();
-  ASSERT_EQ(expectedValue, *locked);
+    constexpr size_t numberOfUseCount = 0;
+    int expectedValue{100};
+    auto s_ptr = cs::shared_ptr<int>(new int{20});
+    auto next_ptr = cs::shared_ptr<int>(new int{expectedValue});
+    cs::weak_ptr<int> weakSomething{s_ptr};
+    weakSomething.reset();
```
Comment (ziobron) [inline homework/shared_ptr/weak_ptr-ut.cpp:68]: you can also check that this reset does not change s_ptr and everything is okay with it. Also, check use_count before reset.

**Review by ziobron — CHANGES_REQUESTED** [review]

File: homework/shared_ptr/control_block.hpp (hunk 9 above)
Comment (Kuba-Wi) [reply homework/shared_ptr/control_block.hpp:32]: I don't want to change so it's const&. It doesn't return pointer because object_ is allocated on stack and in place where I call this method I allocate memory via new. Ok I could return pointer and then dereference it to call new, but it would be pointless.

File: homework/shared_ptr/shared_ptr.hpp (hunk 10 above)
Comment (Kuba-Wi) [reply homework/shared_ptr/shared_ptr.hpp:17]: Previous deleter was faulty.

File: homework/shared_ptr/shared_ptr.hpp (hunk 12 above)
Comment (Kuba-Wi) [reply homework/shared_ptr/shared_ptr.hpp:40]: But one is created on the stack.

File: homework/shared_ptr/shared_ptr-ut.cpp [hunk 18]
```diff
@@ -36,47 +39,53 @@ TEST(sharedPointerTest, MovingContructorShouldMovePointer) {
...
-    ++numberOfUseCount;
-    ASSERT_EQ(copied_ptr.use_count(), numberOfUseCount);
+TEST(sharedPointerTest, resetMethodWithoutArgumentsShouldChangePointedObject) {
+    size_t numberOfUseCount = 2;
+    cs::shared_ptr<int> ptr(new int{0});
+    cs::shared_ptr<int> ptr_doing_nothing(ptr);
+    ASSERT_EQ(ptr.use_count(), numberOfUseCount);
+    ptr.reset();
+    numberOfUseCount = 1;
     ASSERT_EQ(ptr.use_count(), numberOfUseCount);
```
[the remark this answers is not in the document]
Comment (Kuba-Wi) [reply homework/shared_ptr/shared_ptr-ut.cpp]: I checked for std::shared_ptr and use_count before reset was 2.

File: homework/shared_ptr/shared_ptr.hpp (hunk 13 above)
Comment (Kuba-Wi) [reply homework/shared_ptr/shared_ptr.hpp:115]: When I copy shared_ptr that points on nullptr then delete on cb_ is called twice. So more convenient is to leave it as it is and add condition in use_count().

File: homework/shared_ptr/shared_ptr-ut.cpp [hunk 19]
```diff
@@ -36,47 +39,53 @@ TEST(sharedPointerTest, MovingContructorShouldMovePointer) {
...
 }
 
 TEST(makeSharedTest, makeSharedCreatesTheSameValueAsSharedPtr) {
-    cs::shared_ptr<int> someInt(new int{42});
-    auto makeSomeInt = cs::make_shared<int>(42);
+    constexpr int value = 42;
+    cs::shared_ptr<int> someInt(new int{value});
+    auto makeSomeInt = cs::make_shared<int>(value);
     ASSERT_EQ(*makeSomeInt, *someInt);
 }
```
[the remark this answers is not in the document]
Comment (Kuba-Wi) [reply homework/shared_ptr/shared_ptr-ut.cpp:93]: added tests for copying and moving

[the remark this answers is not in the document]
Comment (kuchichan) [reply homework/shared_ptr/shared_ptr-ut.cpp:94]: Also, additional test for `tuple<int, int, int>` 3 parameter and empty string (0 parameter) added.

File: homework/shared_ptr/shared_ptr-ut.cpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (ziobron) [reply homework/shared_ptr/shared_ptr-ut.cpp]: that's true. But my comment was about use_count after reset, not before.
ptr.use_count = 0
ptr_doing_nothing.use_count = 1

**Review by ziobron — APPROVED** [review]
Good job 👍

**Comment by ziobron** [comment]
50 XP za implementację shared_ptra (3794)
25 XP za implementację weak_ptra (3795)
40 XP za implementację make_shared (3796)
{
  "requirement_ids": [3794, 3795, 3796],
   "users": ["kuchichan", "Kuba-Wi"]
}

**Comment by ziobron** [comment]
{
"requirement_ids": [1636],
"users": ["kuchichan", "Kuba-Wi"]
}
